flocker/node/agents/blockdevice.py

```diff
@@ -1308,6 +1460,98 @@ def shared_state_changes(self):
         )
 
 
+# Mapping from desired and discovered dataset state to
+# IStateChange factory. (The factory is expected to take
+# ``desired_dataset`` and ``discovered_dataset``.
+Desired = Discovered = DatasetStates
+DO_NOTHING = lambda **kwargs: NoOp()
+DATASET_TRANSITIONS = {
+    Desired.MOUNTED: {
+        Discovered.NON_EXISTENT:
+            CreateBlockDeviceDataset.from_state_and_config,
+        # Other node will need to deatch first
+        Discovered.ATTACHED_ELSEWHERE: DO_NOTHING,
+        Discovered.ATTACHED_NO_FILESYSTEM:
+            CreateFilesystem.from_state_and_config,
+        Discovered.NON_MANIFEST: AttachVolume.from_state_and_config,
+        DatasetStates.ATTACHED: MountBlockDevice.from_state_and_config,
+    },
+    Desired.NON_MANIFEST: {
+        # Can't create non-manifest datasets yet.
+        Discovered.NON_EXISTENT: DO_NOTHING,
+        # Other node will deatch
+        Discovered.ATTACHED_ELSEWHERE: DO_NOTHING,
+        Discovered.ATTACHED_NO_FILESYSTEM:
+            DetachVolume.from_state_and_config,
+        Discovered.ATTACHED: DetachVolume.from_state_and_config,
+        Discovered.MOUNTED: UnmountBlockDevice.from_state_and_config,
+    },
+    Desired.DELETED: {
+        Discovered.NON_EXISTENT: DO_NOTHING,
+        # Other node will destroy
+        Discovered.ATTACHED_ELSEWHERE: DO_NOTHING,
+        # Can't pick node that will do destruction yet.
+        Discovered.NON_MANIFEST: DO_NOTHING,
+        # This state won't occur until FLOC-1772
+        Discovered.ATTACHED_NO_FILESYSTEM: DetachVolume.from_state_and_config,
+        # This state won't occur until FLOC-1772
+        Discovered.ATTACHED: DetachVolume.from_state_and_config,
+        Discovered.MOUNTED: DestroyBlockDeviceDataset.from_state_and_config,
+    },
+}
```

Comment:
SInce comments above are slightly confusing: DestroyBlockDeviceDataset does unmount-detach-delete, so we need to make sure volume is mounted before deletion will work.